### PR TITLE
Add feature test against buggy V8 versions

### PR DESCRIPTION
**Pull request by sophiebits**
Incorrect property ordering is causing us some problems downstream at
https://github.com/facebook/react/issues/6451.

### CODE REVIEW

**Comment by sindresorhus** [comment]
I'm surprised this is the first time the issue has come up. What do you think about instead just not using the native `Object.assign` at all? The native isn't that much faster than the polyfill and this way you're guaranteed a consistent experience.

**Comment by stevemao** [comment]
> Or maybe expose objectAssign.polyfill() that let's you use only the polyfill directly.

+1, I would like to see this in all your ponyfill modules :)

**Comment by spicyj** [comment]
If I remember right, the native implementation is 2x faster in Chakra (Edge). It's not faster yet in V8 and other engines but there's a chicken-and-egg problem – if it's used commonly in many websites (or popular projects like React), the engine developers have more incentive to optimize it.

You could put the whole shouldUseNative impl in a try/catch block if you're really worried about things breaking strangely. Ultimately if you don't want to take a feature test like this, we should probably switch away from this module in React. I don't mean that as a threat – just that many versions of Node have the broken implementation and we'll need to change something to work around this somewhere in the stack.

**Comment by sindresorhus** [comment]
> You could put the whole shouldUseNative impl in a try/catch block if you're really worried about things breaking strangely.

👍 I'm ok with this change then.

-

Do you think it should be a major or minor release?

-

Btw, I'm happy to add you or anyone on the React Native team to the repo and package in case anything breaks and you need to do a quick fix or something. Let me know.

**Comment by spicyj** [comment]
Updated with try/catch.

I think this should be fine as a minor.

Feel free to add me to the repo and package (also spicyj on npm) – hopefully we won't need any quick fixes but better to be safe. Thanks for the offer.

**Comment by spicyj** [comment]
Okay, going to merge and release a minor.

File: index.js [hunk 1]
```diff
@@ -11,7 +11,51 @@ function toObject(val) {
...
+		}
+
+		// Detect buggy property enumeration order in older V8 versions.
+
+		// https://bugs.chromium.org/p/v8/issues/detail?id=4118
+		var test1 = new String('abc');  // eslint-disable-line
+		test1[5] = 'de';
+		if (Object.getOwnPropertyNames(test1)[0] === '5') {
+			return false;
+		}
```
Comment (jdalton) [inline index.js]: While `Object.getOwnPropertyNames(test1)[0] === '5'` exists in Node v4.0.0 it doesn't effect `Object.assign`.
```js
Object.keys(Object.assign({}, test1))
// => [ '0', '1', '2', '5' ] 
```

File: index.js [hunk 2]
```diff
@@ -11,7 +11,51 @@ function toObject(val) {
...
+		var test2 = {};
+		for (var i = 0; i < 10; i++) {
+			test2['_' + String.fromCharCode(i)] = i;
+		}
+		var order2 = Object.getOwnPropertyNames(test2).map(function (n) {
+			return test2[n];
+		});
+		if (order2.join('') !== '0123456789') {
+			return false;
+		}
```
Comment (jdalton) [inline index.js]: The `order2.join('') !== '0123456789'` bug doesn't effect Node v4.0.0 or above so is not a problem.

File: index.js [hunk 3]
```diff
@@ -11,7 +11,51 @@ function toObject(val) {
...
+
+		// https://bugs.chromium.org/p/v8/issues/detail?id=3056
+		var test3 = {};
+		'abcdefghijklmnopqrst'.split('').forEach(function (letter) {
+			test3[letter] = letter;
+		});
+		if (Object.keys(Object.assign({}, test3)).join('') !==
+				'abcdefghijklmnopqrst') {
+			return false;
+		}
```
Comment (jdalton) [inline index.js]: The `Object.keys(Object.assign({}, test3)).join('') !== 'abcdefghijklmnopqrst'` is the only one that effects Node v4.0.0.

**Pull request merged**